delta_load_atomicity.md

```diff
@@ -0,0 +1,146 @@
+# Graph delta batch load atomicity
+
+## Rollbacks
+
+Note that some of these options do not specifically enable roll backs, but an update can be rolled
+back (unatomically) by:
+
+* Setting any expirations dates equal to `timestamp - 1` to ∞.
+* Deleting any nodes or edges where the `created` field equals `timestamp`, including merge edges.
+* Setting `last_version` to the prior load version on any nodes or edges where `last_version`
+  equals the load version.
+
+If the `timestamp` is in the future at the time the rollback completes, it's as if the update never
+happened. If not, then any queries between the `timestamp` and the completion of the rollback may
+be spurious.
+
+## Option 1: Reject queries
+
+* While an upgrade is in progress reject RE queries.
+  * This is reasonable if updates are quick and can be done at times of low load.
+  * This may cause apps to fail or views to not render.
+  * Note that only queries after `timestamp - 1` need to be rejected while the update is in
+    progress.
+
+## Option 2: Queue queries
+
+* While an upgrade is in progress queue RE queries until the update is complete.
+  * This is reasonable if updates are quick.
+  * This may cause apps or views to wait.
+  * Note that only queries after `timestamp - 1` need to be queued while the update is in
+    progress.
+
+## Option 3: Far future timestamp and faith
+
+* When starting the update, set the `timestamp` to > expected_update_time * 4.
+* The update proceeds as normal.
+* If the update fails, system administrators have until `timestamp` to rollback the upgrade. If
+  they are unsuccessful, queries after `timestamp` may be spurious.
+  * The system could be brought down prior to `timestamp` if necessary.
+
+## Option 4: Restricted timestamps
+* While the update is in progress:
+  * Any query with a timestamp &lt; `timestamp - 2` proceeds normally.
+  * Any other query's behavior depends on an input parameter `strict_timestamp`:
```

Comment:
We could keep a value in the db of "latest_completed_timestamp." We only update it when an update fully completes. Queries for "latest" records automatically use this value. We would have to queue updates to make sure they run serially.

Reply:
The way I was thinking about it was that we wouldn't normally store a value, but when an update started (the timestamp for which is `t0 + t`) we'd ping the API to say that any requests > `t0 + t - 2` should be set to that value. Then we can finish whatever upgrades we want (at long as their timstamps are >=  `t0 + t - 2`) in parallel while the API is resetting timestamps, and when they're all done we remove the timestamp rewrites.

Reply:
But that value should be stored in the DB, yeah